**Functions**


1. What is the difference between a function and a method in Python?
   -A function is a block of reusable code that performs a specific task.

It is defined using the def keyword and can exist independently of classes and objects.

In [ ]:
#Example
def say_hello(name):
    return f"Hello, {name}!"

print(say_hello("saneet"))

Hello, saneet!


In [ ]:
#method call
class Greeter:
    def say_hello(self, name):
        return f"Hello, {name}!"

g = Greeter()
print(g.say_hello("sanjeet"))

Hello, sanjeet!


2.Second question
-Arguments are the actual values passed to the function when it is called function argument.

In [ ]:
#Example
def greet(name):
    print(f"Hello, {name}!")

 Arguments
-Arguments are the actual values passed to the function when it is called.

In [ ]:
greet("sanjeet")

Hello, sanjeet!


Types of Arguments in Python:
Python supports several ways to pass arguments:
Positional arguments – Passed in order:




In [ ]:
def add(a, b):
    return a + b
    add(3,5)

2.Keyword arguments – Passed by name:

In [ ]:
add(a=3, b=5)

8

3.Default parameters – Have default values if not provided:

In [ ]:
def greet(name="Guest"):
    print(f"Hello, {name}!")
    greet()

Variable-length arguments:

args for any number of positional arguments
kwargs for any number of keyword *arguments

In [ ]:
def print_args(*args):
    for arg in args:
        print(arg)

def print_kwargs(**kwargs):
    for key, value in kwargs.items():
        print(key, value)

3. Third question answer
-we define a function using the def keyword and then call it by its name.

In [ ]:
def say_hello():
    print("Hello, World!")

. Function with Return Value:
A function can also return a value that can be used later.

Define the function with a return value:



In [ ]:
def add(a, b):
    return a + b
    #Call the function and use its result:
    result = add(3, 4)

4.forth question answer
-The return statement in a Python function is used to send back a value from the function to the place where it was called. It ends the function’s execution and provides the result that can be used or stored elsewhere in the program.

Purpose of the return statement:
1. Returns a value to the caller.
2. Ends the function immediately when it's called.
3. Allows functions to be used in expressions or assigned to variables.

In [ ]:
#example
def add(a, b):
    return a + b

result = add(5, 3)
print(result)  # Output: 8

8


5. question answer
- iterators and iterables are both used to loop over a sequence of elements, but they are not the same thing.
Iterable:
An iterable is any Python object that can return its elements one at a time.

It must implement the __iter__() method.

Examples: lists, strings, tuples, dictionaries, and sets.

Iterator:
An iterator is an object that keeps track of its current position while iterating.

It implements both __iter__() and __next__() methods.

You can get an iterator from an iterable using the iter() function.

python
Copy code



6. sixth question answer
-generators are a special type of iterator used to generate values on the fly, instead of storing them all at once in memory. They are memory-efficient and often used to work with large data sets or streams.

Generators can be created in two ways:

1. Using a function with the yield keyword:
When a function contains yield instead of return, it becomes a generator.

Each time yield is called, the function pauses and saves its state, then resumes from there when called again.




In [ ]:
#example
def count_up_to(n):
    i = 1
    while i <= n:
        yield i
        i += 1

for number in count_up_to(3):
    print(number)

1
2
3


7. seventh question answer
    -Advantages of Generators:
1. Memory Efficiency
Generators yield one item at a time instead of storing the entire result in memory.

Useful when working with large files, data streams, or infinite sequences.

2. Faster Startup Time
Since values are produced one at a time, the generator starts execution immediately, without waiting to build the whole result.

3. Lazy Evaluation
Values are generated only when needed, which reduces unnecessary computation.

4. Simpler Code
Generators make code easier to write and read, compared to managing state with iterator classes.

5. State Retention
Each time the generator is paused with yield, it remembers its place and resumes from there—great for handling sequences step-by-step.

In [ ]:
#example
def get_numbers():
    return [i for i in range(1000000)]  # stores all numbers in memory


8. eighth question answer
  -A lambda function in Python is a small, anonymous function that is defined using the lambda keyword instead of the regular def keyword. It is used for creating short, one-line functions without giving them a name.

  When is it typically used?
Lambda functions are commonly used when:

You need a short function temporarily.

You’re passing a function as an argument (e.g., in map(), filter(), or sorted()).

You want cleaner, more compact code.



In [ ]:
# Sort a list of tuples by the second value
pairs = [(1, 2), (3, 1), (5, 0)]
pairs.sort(key=lambda x: x[1])
print(pairs)  # Output: [(5, 0), (3, 1), (1, 2)]


[(5, 0), (3, 1), (1, 2)]


9. nineth question answer
  -The map() function in Python is used to apply a function to every item in an iterable (like a list or tuple) and returns a new iterable (a map object) with the results.
  Purpose:
To transform each element in a sequence by applying a given function, without writing a loop manually.
map(function, iterable)
function: a function to apply (can be a regular or lambda function)
iterable: a sequence (like list, tuple, etc.)




In [ ]:
#example
def square(x):
    return x * x

numbers = [1, 2, 3, 4]
result = map(square, numbers)
print(list(result))  # Output: [1, 4, 9, 16]


[1, 4, 9, 16]


map() returns a map object (which is an iterator), so you often wrap it with list() or tuple() to see the results.

It’s useful when you want to apply the same operation to each item in a collection.



10. tenth question answer
  -map(), reduce(), and filter() are functional programming tools used to process data in iterables like lists and tuples. Each has a different purpose:
  🔹 map(function, iterable)
Purpose: Applies a function to each item in an iterable.

Returns: A new iterable with transformed items.



In [ ]:
numbers = [1, 2, 3]
result = map(lambda x: x * 2, numbers)
print(list(result))  # Output: [2, 4, 6]


[2, 4, 6]


🔹 filter(function, iterable)
Purpose: Filters out items in an iterable for which the function returns True.

Returns: A new iterable with only the filtered items.

In [ ]:
numbers = [1, 2, 3, 4]
result = filter(lambda x: x % 2 == 0, numbers)
print(list(result))  # Output: [2, 4]


[2, 4]


🔹 reduce(function, iterable)
Purpose: Repeatedly applies a function to the items of an iterable, reducing it to a single value.

Returns: A single result.

Must be imported from functools.

In [ ]:
from functools import reduce

numbers = [1, 2, 3, 4]
result = reduce(lambda x, y: x + y, numbers)
print(result)  # Output: 10


10


Use **map** to modify elements,

Use **filter** to select elements,

Use **reduce** to combine elements.

11. eleventh question answer
  -Given:
List = [47, 11, 42, 13]

We want to sum all elements using reduce().

In [ ]:
from functools import reduce
reduce(lambda x, y: x + y, [47, 11, 42, 13])


113

1. Step-by-step (internal working):
Start with the first two elements:
Step 1: x = 47, y = 11 → x + y = 58
2. Take the result (58) and apply the function to the next element:
Copy code
Step 2: x = 58, y = 42 → x + y = 100
3. Again, take the result (100) and apply it to the next element:
Copy code
Step 3: x = 100, y = 13 → x + y = 113

**practical Question**
1. first question answer
  - Python function that takes a list of numbers and returns the sum of all even numbers in the list:

In [46]:
def sum_even_numbers(numbers):
    return sum(num for num in numbers if num % 2 == 0)

# Example usage:
my_list = [1, 2, 3, 4, 5, 6]
result = sum_even_numbers(my_list)
print("Sum of even numbers:", result)


Sum of even numbers: 12


Here's a Python function that takes a list of numbers and returns the sum of all even numbers in the list:

```python
def sum_even_numbers(numbers):
    return sum(num for num in numbers if num % 2 == 0)

# Example usage:
my_list = [1, 2, 3, 4, 5, 6]
result = sum_even_numbers(my_list)
print("Sum of even numbers:", result)
```

### How it works:

* The function uses a generator expression to filter and sum only the numbers that are divisible by 2 (even numbers).
* `num % 2 == 0` checks for evenness.

Let me know if you’d like a version that handles non-integer values or includes input validation.


2. second question answer
    - Python function that takes a string and returns its reverse:

In [47]:
def reverse_string(s):
    return s[::-1]

# Example usage:
input_str = "hello"
reversed_str = reverse_string(input_str)
print("Reversed string:", reversed_str)


Reversed string: olleh


**Explanation:**
s[::-1] is a slicing technique that reverses the string by stepping through it backwards.

Let me know if you'd like this to be case-insensitive, remove spaces, or handle special characters differently.

3. Third question answer
    - Python function that takes a list of integers and returns a new list with the squares of each number:

In [48]:
def square_numbers(numbers):
    return [num ** 2 for num in numbers]

# Example usage:
input_list = [1, 2, 3, 4, 5]
squared_list = square_numbers(input_list)
print("Squared numbers:", squared_list)


Squared numbers: [1, 4, 9, 16, 25]


This uses a list comprehension to create a new list where each element is the square (num ** 2) of the corresponding element in the input list.

Let me know if you want to filter negative numbers or handle input validation!

4. Forth question answer.
    - A Python function that checks whether a given number between 1 and 200 is prime:

In [49]:
def is_prime(n):
    if n < 2 or n > 200:
        return False  # Not in the range 1 to 200 or less than 2 (not prime)
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True

# Example usage:
for num in range(1, 201):
    if is_prime(num):
        print(f"{num} is a prime number.")


2 is a prime number.
3 is a prime number.
5 is a prime number.
7 is a prime number.
11 is a prime number.
13 is a prime number.
17 is a prime number.
19 is a prime number.
23 is a prime number.
29 is a prime number.
31 is a prime number.
37 is a prime number.
41 is a prime number.
43 is a prime number.
47 is a prime number.
53 is a prime number.
59 is a prime number.
61 is a prime number.
67 is a prime number.
71 is a prime number.
73 is a prime number.
79 is a prime number.
83 is a prime number.
89 is a prime number.
97 is a prime number.
101 is a prime number.
103 is a prime number.
107 is a prime number.
109 is a prime number.
113 is a prime number.
127 is a prime number.
131 is a prime number.
137 is a prime number.
139 is a prime number.
149 is a prime number.
151 is a prime number.
157 is a prime number.
163 is a prime number.
167 is a prime number.
173 is a prime number.
179 is a prime number.
181 is a prime number.
191 is a prime number.
193 is a prime number.
197 is a prime nu

Explanation:
The function returns False for numbers less than 2 or greater than 200.

It checks divisibility from 2 up to the square root of n for efficiency.

If no divisors are found, the number is prime.

Let me know if you'd like a version that returns all prime numbers in the range instead!

5. fifth question answer
      -A Python iterator class that generates the Fibonacci sequence up to a specified number of terms:   

In [50]:
class FibonacciIterator:
    def __init__(self, max_terms):
        self.max_terms = max_terms
        self.count = 0
        self.a = 0
        self.b = 1

    def __iter__(self):
        return self

    def __next__(self):
        if self.count >= self.max_terms:
            raise StopIteration
        if self.count == 0:
            self.count += 1
            return 0
        elif self.count == 1:
            self.count += 1
            return 1
        else:
            fib = self.a + self.b
            self.a, self.b = self.b, fib
            self.count += 1
            return fib

# Example usage:
fib_seq = FibonacciIterator(10)
for num in fib_seq:
    print(num)


0
1
1
2
3
5
8
13
21
34


Explanation:
The class tracks the number of terms generated using self.count.

It maintains the last two Fibonacci numbers (self.a and self.b).

__iter__() returns the iterator object itself.

__next__() generates the next Fibonacci number until the specified limit.

Let me know if you'd like a version that generates Fibonacci numbers up to a maximum value instead of a fixed number of terms.

6. sixth question answer
    - a Python generator function that yields the powers of 2 up to a given exponent:

In [51]:
def powers_of_two(max_exponent):
    for exponent in range(max_exponent + 1):
        yield 2 ** exponent

# Example usage:
for power in powers_of_two(5):
    print(power)


1
2
4
8
16
32


Explanation:
The generator uses a for loop to iterate from 0 to max_exponent.

yield pauses the function and returns the current power of 2.

2 ** exponent calculates the current power of 2.

Output for powers_of_two(5) would be:

Copy code
1
2
4
8
16
32
Let me know if you want the results returned as a list or with additional formatting.


7. seventh question answer.
      -A Python generator function that reads a file line by line and yields each line as a string:

In [ ]:
def read_file_lines(filepath):
    with open(filepath, 'r') as file:
        for line in file:
            yield line.rstrip('\n')  # Remove trailing newline character

# Example usage:
# Replace 'example.txt' with the path to your file
for line in read_file_lines('example.txt'):
    print(line)


Explanation:
with open(...) ensures the file is properly closed after reading.

The for line in file loop reads one line at a time (memory efficient).

yield returns each line to the caller.

.rstrip('\n') strips the newline character from the end of each line.

Let me know if you want to include empty lines, strip whitespace, or handle large files differently!


8.eighth question answer
  -A lambda function in Python to sort a list of tuples based on the second element of each tuple:

In [53]:
# Example list of tuples
tuples_list = [(1, 3), (4, 1), (2, 2), (5, 0)]

# Sort based on the second element using a lambda function
sorted_list = sorted(tuples_list, key=lambda x: x[1])

print("Sorted list:", sorted_list)


Sorted list: [(5, 0), (4, 1), (2, 2), (1, 3)]


9. nineth question answer.
      -A Python program that uses the map() function to convert a list of temperatures from Celsius to Fahrenheit:

In [54]:
# Function to convert Celsius to Fahrenheit
def celsius_to_fahrenheit(celsius):
    return (celsius * 9/5) + 32

# List of temperatures in Celsius
celsius_temps = [0, 20, 37, 100, -10]

# Use map() to apply the conversion function to each item in the list
fahrenheit_temps = list(map(celsius_to_fahrenheit, celsius_temps))

# Print the result
print("Celsius:", celsius_temps)
print("Fahrenheit:", fahrenheit_temps)


Celsius: [0, 20, 37, 100, -10]
Fahrenheit: [32.0, 68.0, 98.6, 212.0, 14.0]


10 tenth question answer.
      -Python program that uses the filter() function to remove all vowels from a given string:

In [55]:
# Input string
input_str = "Hello, this is a sample sentence."

# Define a function that returns True if the character is not a vowel
def is_not_vowel(char):
    return char.lower() not in 'aeiou'

# Use filter() to remove vowels
no_vowels = ''.join(filter(is_not_vowel, input_str))

# Print the result
print("Original String: ", input_str)
print("Without Vowels:  ", no_vowels)


Original String:  Hello, this is a sample sentence.
Without Vowels:   Hll, ths s  smpl sntnc.


11 eleventh question answer.
    - Python Program using lambda and map

In [56]:
# Sample list of orders
orders = [
    [34587, "Learning Python", 1, 40.95],
    [98762, "Programming Python", 5, 56.80],
    [77226, "Head First Python", 3, 32.95],
    [88112, "Einführung in Python", 3, 24.99]
]

# Use map and lambda to compute final prices with €10 fee if under €100
final_orders = list(map(
    lambda order: (
        order[0],
        round(order[2] * order[3] + (10 if order[2] * order[3] < 100 else 0), 2)
    ),
    orders
))

# Output the result
print(final_orders)


[(34587, 50.95), (98762, 284.0), (77226, 108.85), (88112, 84.97)]
